linalg/solvers.cpp

```diff
@@ -1319,6 +1319,54 @@ void NewtonSolver::Mult(const Vector &b, Vector &x) const
    final_norm = norm;
 }
 
+void NewtonSolver::CheckJacobian(Vector &x, Array<int> &ess_tdof)
+{
+   MFEM_ASSERT(oper != nullptr, "Call SetOperator first.");
+
+   double eps = 1e-8;
+   Vector xpert(height), one(height), fx(height),
+          fxp(height), basis(height), jac_col(height);
+
+   basis = 0.0;
+   xpert = x;
+
+   Operator &jac = oper->GetGradient(x);
+
+   oper->Mult(x, fx);
+
+   for (int j = 0; j < height; j++)
+   {
+      // Is column associated with an essential boundary?
+      if (ess_tdof.FindSorted(j) >= 0)
+      {
+         jac_col = 0.0;
+      }
+      else
+      {
+         basis(j) = 1.0;
+
+         // Forward finite differences
+         xpert(j) += eps;
```

Comment:
It looks like you're forming a finite difference approximation $Df(x)e_{j} = \frac{f(x + h) - f(x)}{\|h\|}$ for some appropriate norm. If I pass in a `GridFunction gf` associated with, for example, an element of $H^{1}$, into `CheckJacobian` as its vector argument `x`, it seems like this code assumes in lines 1349 to 1352 that the norm of `xpert - x` is `eps`. Is that assumption true? It looks like everything is correct if the object passed by reference into `Vector& x` is not associated with a `FiniteElementSpace` -- as I understand it, this would correspond to `x` being an element of $\mathbb{R}^{n}$ for some $n$.